### ISSUE TITLE
borg2: symlink source vs. target

### ISSUE THREAD

**Issue by ThomasWaldmann** [borgbackup/borg#7245]
there seem to be different naming conventions about symlink source and target.

Let's consider the result of `ln -s A B`.

Convention1: Some say A is the `source` and `B` is the target (which seems a valid PoV, especially when describing the **creation of a new symlink**, because A exists and B will be created, similar to `cp A B`).

Convention2: But when later looking at an **existing symlink**, I guess it is more popular to say that A is the `symlink target` and B is the symlink.

When borg1 archives a symlink, it creates an archive item like convention1:
```
item.path = B
item.source = A
```

But I somehow feel that convention2 is more popular/natural when talking about symlinks, so borg2 could change it to be:
```
item.path = B
item.target = A
```

Existing borg1 archives would be converted on the fly to the borg2 format when using `borg transfer`.

**Comment by ThomasWaldmann** [comment]
Like this post, if you think convention1 (like borg1) should stay as is.

**Comment by ThomasWaldmann** [comment]
Like this post if you think we should change to convention2 for borg2.

**Comment by lfam** [comment]
IMO, convention 1 is explicitly incorrect. The manpage for coreutils `ln(1)` starts like this:

`ln [OPTION]... [-T] TARGET LINK_NAME`

So, the pre-existing file is the target, and the other component is the symbolic link to the target.

It's the same for `link(2)`, the syscall.

In general, there is a lot of confusion among users about which is which, and how to describe them. But, I think the best course of action is to stay close to the terminology used by coreutils, which can be considered canonical for Linux systems. So, I would support your suggestion for borg 2:

```
item.path = B
item.target = A
item.mode = <IS_SYMLINK>
m```

**Comment by ThomasWaldmann** [comment]
@lfam I have seen man pages with SOURCE at the place of TARGET.

Also, when you think about the close relation of `ln -s` to `cp` (which is also a good way to remember order in the `ln -s` command), that even makes sense for describing the **action of creating a symlink/copy**.

**Comment by lfam** [comment]
Interesting, I haven't seen man pages like that, although I only use Linux with coreutils.

**Comment by klemensn** [comment]
> Interesting, I haven't seen man pages like that, although I only use Linux with coreutils.

Busybox ln(1) seems to follow GNU coreutils ln(1) .

Here are systems doing it the other way:
[4.3BSD ln(1) from 1986](https://minnie.tuhs.org/cgi-bin/utree.pl?file=4.3BSD/usr/man/man1/ln.1)
```
SYNOPSIS
    ln [-s] sourcename [targetname]
    ln [-s] sourcename1 sourcename2 [ sourcename3 ... ] targetdirectory
```

Same source/target semantic today on NetBSD, FreeBSD, OpenBSD and Solaris (with varying wording), e.g. OpenBSD [ln(1)](https://man.openbsd.org/ln.1):
```
$ man -h ln
ln [-fhLnPs] source [target]
ln [-fLPs] source ... [directory]
```

Conversely, [readlink(1)](https://man.openbsd.org/readlink.1) says
```
DESCRIPTION
     The readlink utility when invoked with the pathname of a symbolic link as
     its argument dereferences the symbolic link and prints the name of target
     on standard output. [...]
```

**Comment by lfam** [comment]
@ThomasWaldmann 

I currently have access to coreutils 8.32, which is not the latest. In that man page

**Comment by lfam** [comment]
Wow, I wouldn't have guessed that BSD would be backwards from GNU. But okay!

**Comment by cherti** [comment]
I think it boils down to the exact command. If the command is about _linking something_, I would expect `thing symlink`, if it's about _creating a link_, I would expect `symlink thing`. The thing it's about comes first, the parameter comes second, now the question is whether it's about linking something or creating a link.

For `ln`, as it stands for "link", in this reading `thing symlink` is semantically correct, I'd argue.

**Comment by goebbe** [comment]
I believe the term "source" is ambiguous/ confusing and not self-explanatory in the context of symlinks.  The latest manpage of coreutils avoids this confusion.

**Comment by cherti** [comment]
Then I would go with `path` and `target`, because that's closest to the terminology the manpages use. (and `link_name` instead of `path` might clash with borg's internal structure, from how I understand that snippet.)

**Comment by ThomasWaldmann** [comment]
@cherti yes, `item.path` always refers to the full path of the filesystem item backed up in that data structure (no matter what type the item is), so this should not and can not be changed.

**Comment by ThomasWaldmann** [comment]
OK, this was pretty close, but I will change it to `target` for borg2 (convention2).
